### PR TITLE
WIP - Added first round of Windows modules

### PR DESCRIPTION
**Pull request by jborean93**
First round proposal of the Windows modules that should stay with Ansible for a stdlib/minimal implementation. I tried to find a win_ variant for each of the POSIX modules included here but it also adds a few useful Windows specific modules as well.

### CODE REVIEW

File: scenarios/minimal/ansible.yml [hunk 1]
```diff
@@ -22,7 +22,12 @@ _core:
     - uri.py
     - yum.py
     - wait_for_connection.py
+    - win_copy
+    - win_reboot
+    - win_template
+    # - win_updates - not sure on this one
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: This is a pretty complex action plugin/module and intertwines with the become process which might be fragile if it was to move away from the codebase. Also update are important in Windows so I would lean towards adding this in.

File: scenarios/minimal/ansible.yml [hunk 2]
```diff
@@ -50,7 +56,10 @@ _core:
     - inventory_cache.py
     - return_common.py
     - shell_common.py
+    - shell_windows.py
+    - template_common.py
     - url.py
+    - url_windows.py
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: Needed for `win_uri` and `win_get_url`

File: scenarios/minimal/ansible.yml [hunk 3]
```diff
@@ -50,7 +56,10 @@ _core:
     - inventory_cache.py
     - return_common.py
     - shell_common.py
+    - shell_windows.py
+    - template_common.py
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: Needed for `template` and `win_template`

File: scenarios/minimal/ansible.yml [hunk 4]
```diff
@@ -136,7 +147,8 @@ _core:
     - files/lineinfile.py
     - files/patch.py
     - files/replace.py
-    - files/stat.py
+    - files/stat.pyi
+    - files/tempfile.py
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: Uses in the `setup_remote_tmp_dir` in `ansible-test` which I think we should have by default

File: scenarios/minimal/ansible.yml [hunk 5]
```diff
@@ -193,7 +205,71 @@ _core:
     - utilities/logic/set_stats.py
     - utilities/logic/wait_for.py
     - utilities/logic/wait_for_connection.py
+    - windows/async_status.ps1
+    - windows/setup.ps1
+    - windows/slurp.ps1
+    # - windows/win_acl.ps1
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: The acl/acl_inheritance/owner modules all deal with permissions on a file/reg key. I'm not the biggest fan of how these are set up right now but have no alternative. Leaving them here as file permissions are pretty important on Windows.

File: scenarios/minimal/ansible.yml [hunk 6]
```diff
@@ -193,7 +205,71 @@ _core:
...
+    # - windows/win_acl.py
+    # - windows/win_acl_inheritance.ps1
+    # - windows/win_acl_inheritance.py
+    - windows/win_command.ps1
+    - windows/win_command.py
+    - windows/win_copy.ps1
+    - windows/win_copy.py
+    # - windows/win_dsc.ps1
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: Pretty big all rounder and opens up Ansible functionality out of the box quite a lot. Technically doesn't need to be here but is quite useful.

File: scenarios/minimal/ansible.yml [hunk 7]
```diff
@@ -193,7 +205,71 @@ _core:
...
+    # - windows/win_dsc.py
+    # - windows/win_environment.ps1
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: Sets environment variables for Windows, Linux can be done with `lineinfile` but Windows really need this.

File: scenarios/minimal/ansible.yml [hunk 8]
```diff
@@ -193,7 +205,71 @@ _core:
...
+    # - windows/win_environment.py
+    - windows/win_feature.ps1
+    - windows/win_feature.py
+    - windows/win_find.ps1
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: Technically community but there is `find.py` so I added this here

File: scenarios/minimal/ansible.yml [hunk 9]
```diff
@@ -193,7 +205,71 @@ _core:
...
+    - windows/win_file.py
+    - windows/win_get_url.ps1
+    - windows/win_get_url.py
+    - windows/win_group.ps1
+    - windows/win_group.py
+    - windows/win_group_membership.ps1
+    - windows/win_group_membership.py
+    - windows/win_hostname.ps1
+    - windows/win_hostname.py
+    - windows/win_lineinfile.ps1
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: Only added due to `lineinfile.py` being present, not sure if we want this for Windows.

File: scenarios/minimal/ansible.yml [hunk 10]
```diff
@@ -193,7 +205,71 @@ _core:
...
+    - windows/win_lineinfile.py
+    # - windows/win_owner.ps1
+    # - windows/win_owner.py
+    - windows/win_package.ps1
+    - windows/win_package.py
+    # - windows/win_path.ps1
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: Same as `win_environment`, just more helpful for path based env var.

File: scenarios/minimal/ansible.yml [hunk 11]
```diff
@@ -193,7 +205,71 @@ _core:
...
+    # - windows/win_path.py
+    - windows/win_ping.ps1
+    - windows/win_ping.py
+    - windows/win_reboot.py
+    - windows/win_reg_stat.ps1
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: Registry is quite important in Windows, the `win_reg_stat` and `win_regedit` modules should be present.

File: scenarios/minimal/ansible.yml [hunk 12]
```diff
@@ -193,7 +205,71 @@ _core:
...
+    - windows/win_reg_stat.py
+    - windows/win_regedit.ps1
+    - windows/win_regedit.py
+    - windows/win_service.ps1
+    - windows/win_service.py
+    - windows/win_shell.ps1
+    - windows/win_shell.py
+    - windows/win_stat.ps1
+    - windows/win_stat.py
+    - windows/win_tempfile.ps1
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: Same as `tempfile.py`, really useful for ansible-test

File: scenarios/minimal/ansible.yml [hunk 13]
```diff
@@ -193,7 +205,71 @@ _core:
...
+    - windows/win_tempfile.py
+    - windows/win_template.py
+    - windows/win_timezone.ps1
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: I question this one, only here because `timezone.py` is also present

File: scenarios/minimal/ansible.yml [hunk 14]
```diff
@@ -193,7 +205,71 @@ _core:
...
+    # - windows/win_updates.ps1
+    # - windows/win_updates.py
+    - windows/win_uri.ps1
+    - windows/win_uri.py
+    - windows/win_user.ps1
+    - windows/win_user.py
+    - windows/win_wait_for.ps1
+    - windows/win_wait_for.py
     shell:
+    - cmd.py
```
Comment (jborean93) [inline scenarios/minimal/ansible.yml]: Required for using ssh on Windows with the default of Win32 OpenSSH

**Comment by jborean93** [comment]
@gundalow that's part of the issue, I'm still not 100% clear as to what defines minimal. It seems like everbody and their dog have their own opinions. The current readme states that 

> Opinionated 'minimal' usable Ansible version, keeps certain things in Core that are either required for any Ansible cli to execute or allow minimal configuration of the system to then enable installing collections and/or other content.

This is different from what I've heard from others. Other POV I've heard is that minimal is actually stdlib where it includes just the basic batteries included work to get going for very basic scenarios and this is what that PR is meant to reflect. It is designed to mirror the same functionality (with a few Windows specific extras) from the POSIX side as well as include things which would be nice to have for ansible-test to include by default in it's testing. This is more along the lines with my thoughts on what a stdlib actually is.

**Comment by jillr** [comment]
@jborean93 Several AWS tests depend on `win_file`, is this scenario still a reasonable example to build a collection to test against?

**Comment by abadger** [comment]
At this weeks meeting, this was discussed and the direction is that controller-side content will be included but module-side will be in a separate collection.  I'll open a pull request based on that and CC @jborean93 to look more closely at it.

**Comment by abadger** [comment]
PR to move the windows connection and shell plugins: https://github.com/ansible-community/collection_migration/pull/258

I'm leaving this open as it might be the basis of deciding what goes in to a supported windows collection versus a community windows collection.

**Comment by jborean93** [comment]
This lives again in https://github.com/ansible-community/collection_migration/pull/377

**Pull request closed without merging**